lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs

```diff
@@ -0,0 +1,787 @@
+    ( TokenBundle (..) )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( TokenMap )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxConstraints (..)
+    , txOutputCoinCost
+    , txOutputHasValidSize
+    , txOutputHasValidTokenQuantities
+    )
+import Control.Monad
+    ( (>=>) )
+import Data.Bifunctor
+    ( first )
+import Data.Either.Extra
+    ( eitherToMaybe, maybeToEither )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Maybe
+    ( catMaybes, listToMaybe )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
+import qualified Data.Foldable as F
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Set as Set
+
+--------------------------------------------------------------------------------
+-- Selections
+--------------------------------------------------------------------------------
+
+data Selection i s = Selection
+    { inputIds :: !(NonEmpty i)
+      -- ^ The selected inputs.
+    , inputBalance :: !TokenBundle
+      -- ^ The total balance of value provided by the inputs.
+    , outputs :: !(NonEmpty TokenBundle)
+      -- ^ The outputs, adjusted to pay for the fee.
+    , fee :: !Coin
+      -- ^ The actual fee payable for this selection.
+    , feeExcess :: !Coin
+      -- ^ The excess over the minimum permissible fee for this selection.
+    , size :: !s
+      -- ^ The size of this selection.
+    , rewardWithdrawal :: !Coin
+      -- ^ The reward withdrawal amount, if any.
+    }
+    deriving (Eq, Generic, Show)
+
+newtype RewardWithdrawal = RewardWithdrawal
+    { unRewardWithdrawal :: Coin }
+
+--------------------------------------------------------------------------------
+-- Selection correctness
+--------------------------------------------------------------------------------
+
+data SelectionCorrectness s
+    = SelectionCorrect
+    | SelectionAssetBalanceIncorrect
+      SelectionAssetBalanceIncorrectError
+    | SelectionFeeIncorrect
+      SelectionFeeIncorrectError
+    | SelectionFeeExcessIncorrect
+      SelectionFeeExcessIncorrectError
+    | SelectionFeeInsufficient
+      SelectionFeeInsufficientError
+    | SelectionOutputBelowMinimumAdaQuantity
+      SelectionOutputBelowMinimumAdaQuantityError
+    | SelectionOutputSizeExceedsLimit
+      SelectionOutputSizeExceedsLimitError
+    | SelectionSizeExceedsLimit
+     (SelectionSizeExceedsLimitError s)
+    | SelectionSizeIncorrect
+     (SelectionSizeIncorrectError s)
+    deriving (Eq, Show)
+
+verify
+    :: TxSize s
+    => TxConstraints s
+    -> Selection i s
+    -> SelectionCorrectness s
+verify constraints selection
+    | Just e <- checkAssetBalance selection =
```

Comment:
Haven't seen this style before. Wouldn't 
```haskell
checkAssetBalance :: Selection i s -> Either SelectionAssetBalanceIncorrectError ()
```
and
```haskell
verify = do
    left SelectionAssetBalanceIncorrect $ checkAssetBalance selection
    left SelectionFeeIncorrect $ ...
    ...
```
be more plain?

Reply:
I agree that the control flow could be a little clearer.

What do you think about this solution: d31d1ec5a72202561fb4f15012ec6ba1dbf1d229?

Reply:
I think it reads better, but I don't see why we need `SelectionCorrectness` https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r617352652